### ISSUE TITLE
A Segmentation fault when multi-property check

### ISSUE THREAD

**Issue by Anthonysdu** [esbmc/esbmc#1408]
```
#include <assert.h>
#include <string.h>
#include <stdlib.h>
struct InnerStruct {
    char innerValue[10];
};

int main() {
    struct InnerStruct array[2];

    strcpy(array[0].innerValue, "20");
    strcpy(array[1].innerValue, "40");

    for (int i = 0; i < 2; i++) {
        int innerValue = atoi(array[i].innerValue);
        assert(innerValue == 40);
    }

    return 0;
}
```

`esbmc main.c --multi-property --unwind 3 --no-unwinding-assertions` or `esbmc main.c --multi-property --unwind 3`

```
Solving claim 'assertion innerValue == 40' with solver Boolector 3.2.1
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Fetching array elements inside tuples currently unimplemented, returning empty expression...
Segmentation fault (core dumped)
```

`esbmc main.c --unwind 3
`
```
State 18 file main.c line 15 column 9 function main thread 0
----------------------------------------------------
  innerValue = 20 (00000000 00000000 00000000 00010100)

State 19 file main.c line 16 column 9 function main thread 0
----------------------------------------------------
Violated property:
  file main.c line 16 column 9 function main
  assertion innerValue == 40
  innerValue == 40


VERIFICATION FAILED
```

**Comment by lucasccordeiro** [comment]
@ChenfengWei0: could you please prioritize this issue?

**Comment by ChenfengWei0** [comment]
The bug is caused by

```c++
build_goto_trace(local_eq, runtime_solver, goto_trace, is_compact_trace);
```
- if `is_compact_trace == false`, then it will trigger this bug
- if `is_compact_trace == true`, then it runs correctly

Actually, we do not need to `multi-property` to trigger this bug.

**example**
```c++
struct InnerStruct
{
   char innerValue[10];
};

int main()
{
   struct InnerStruct array[2];
   strcpy(array[0].innerValue, "20");
   char *ptr;
   assert(strtol(array[0].innerValue, ptr, 10) == 40);
}
```

- `--unwind 3 --no-pointer-check --no-bound --no-div-by-zero-check`
```
[Counterexample]


State 1 file 2.c line 1218 column 4 function main thread 0
----------------------------------------------------
Violated property:
  file 2.c line 1218 column 4 function main
  assertion strtol(array[0].innerValue, ptr, 10) == 40
  return_value$_strtol$1 == 40


VERIFICATION FAILED
```

- `--unwind 3 --no-pointer-check --no-bound --no-div-by-zero-check --no-slice`
```
[PROGRESS] Building error trace
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
[1]    20474 segmentation fault  ./esbmc 2.c --unwind 3 --color --no-pointer-check --no-bound  --no-slice
```

**Comment by fbrausse** [comment]
The bug is a segfault in `smt_convt::get()`. It is being called on a nil sub-expr which is the result of `smt_tuple_node_flattener::tuple_get_array_elem()`, as also logged:
```
[WARNING] Fetching array elements inside tuples currently unimplemented, returning empty expression...
```
This patch avoids calling `get()` on empty subexprs and fixes the examples:
```diff
diff --git a/src/solvers/smt/smt_conv.cpp b/src/solvers/smt/smt_conv.cpp
index 00fce31695..85ac72f7d7 100644
--- a/src/solvers/smt/smt_conv.cpp
+++ b/src/solvers/smt/smt_conv.cpp
@@ -2363,7 +2363,9 @@ expr2tc smt_convt::get(const expr2tc &expr)
   // Recurse on operands
   bool have_all = true;
   res->Foreach_operand([this, &have_all](expr2tc &e) {
-    expr2tc new_e = get(e);
+    expr2tc new_e;
+    if(e)
+      new_e = get(e);
     e = new_e;
     if(!e)
       have_all = false;
```
I actually thought this was fixed in 9ec18541878 already, but appearently not fully. We may need to find a better approach as the above patch still could let nil exprs through when they are two or more sub-expr levels deep.

@rafaelsamenezes I remember we talked about this earlier, and I believe the suggestion back then was to let `get()` also return the `have_all`. Did this ever get implemented somewhere?

**Comment by rafaelsamenezes** [comment]
> @rafaelsamenezes I remember we talked about this earlier, and I believe the suggestion back then was to let get() also return the have_all. Did this ever get implemented somewhere?

Hm, I was playing with it but never finished an actual patch.

Nevertheless, it would be good to clarify if the multi property crash is due to some other VCC failing.
